bundles/org.openhab.binding.irobot/src/main/java/org/openhab/binding/irobot/handler/RoombaHandler.java

```diff
@@ -0,0 +1,520 @@
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.irobot.internal.IdentProtocol;
+import org.openhab.binding.irobot.internal.IdentProtocol.IdentData;
+import org.openhab.binding.irobot.internal.MQTTProtocol;
+import org.openhab.binding.irobot.internal.RawMQTT;
+import org.openhab.binding.irobot.roomba.RoombaConfiguration;
+import org.openhab.binding.irobot.roomba.RoombaMqttBrokerConnection;
+import org.openhab.core.config.core.Configuration;
+import org.openhab.core.library.types.DecimalType;
+import org.openhab.core.library.types.OnOffType;
+import org.openhab.core.library.types.StringType;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.binding.BaseThingHandler;
+import org.openhab.core.types.Command;
+import org.openhab.core.types.RefreshType;
+import org.openhab.core.types.State;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonParseException;
+import com.google.gson.JsonPrimitive;
+import com.google.gson.stream.JsonReader;
+
+/**
+ * The {@link RoombaHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author hkuhn42 - Initial contribution
+ * @author Pavel Fedin <pavel_fedin@mail.ru> - Rewrite for 900 series
+ */
+@NonNullByDefault
+public class RoombaHandler extends BaseThingHandler {
+    private final Logger logger = LoggerFactory.getLogger(RoombaHandler.class);
+    private final ExecutorService singleThread = Executors.newSingleThreadExecutor();
+    private static final int reconnectDelay = 5; // In seconds
+    private @Nullable Future<?> reconnectReq;
+    // Dummy RoombaConfiguration object in order to shut up Eclipse warnings
+    // The real one is set in initialize()
+    private RoombaConfiguration config = new RoombaConfiguration();
+    private @Nullable String blid = null;
+    private @Nullable RoombaMqttBrokerConnection connection;
+    private Hashtable<String, State> lastState = new Hashtable<String, State>();
+    private MQTTProtocol.@Nullable Schedule lastSchedule = null;
+    private boolean auto_passes = true;
+    private @Nullable Boolean two_passes = null;
+    private boolean carpet_boost = true;
+    private @Nullable Boolean vac_high = null;
+    private boolean isPaused = false;
+
+    public RoombaHandler(Thing thing) {
+        super(thing);
+    }
+
+    @Override
+    public void initialize() {
+        logger.trace("initialize()");
+        config = getConfigAs(RoombaConfiguration.class);
+        updateStatus(ThingStatus.UNKNOWN);
+        connect();
+    }
+
+    @Override
+    public void dispose() {
+        logger.trace("dispose()");
+
+        singleThread.execute(() -> {
```

Comment:
In order to run in the same thread where connect requests are handled, in order to avoid race conditions. Is it OK to call connection.stop() while connection.start() is running? I suggest the MQTT client would badly screw up
Perhaps Java has a different view on it, then teach me please. I come from C world, and there this would be the right way to do things, if you don't know where initial requests come from.

Reply:
How comes? They are called only from within singleThreadExecutor. It's supposed to be single thread, isn't it ?

Reply:
You're right, but spawning a thread only to achieve synchronization is a bit unreasonable. Bindings should only create new threads if necessary. In this case you could simply add the `synchronized` modifier to the methods you want to synchronize.

Reply:
As far as i understand, scheduler provided by the base class does the same, but it can run more than one thread if it feels up to, as it doesn't guarantee any ordering of scheduled tasks. But OK, as you wish, here's your synchronized, let's learn some java :)